content/posts/cicd_ha.md

```diff
@@ -0,0 +1,132 @@
+* **Decouple also the orchestration layer**. Smaller and more masters (orchestrators) are going to be more scalable and resilient to any overall failure. If one master fails it is not impacting any other orchestrator and its pipelines. So platform availability improves with scalability.
+* Rely on a **native self-healing, resilient and scalable infrastructure**. Event thought replication is not needed, we need to recover fast from any hardware or service downtime.
+* **Decouple also pipeline execution**. Using different agent executors during pipeline execution improves availability, just the same way that *assembly lines* in the industry do.
+
+It can be pretty cool if I can access my jobs or builds all the time because my masters are all the time up and running, but it is not that cool if they are not producing the same results in terms of building and delivering software because my pipelines are not executing and recovering on time. Again... resilience first.
+
+## Scalability and Kubernetes as HA best practice
+
+If we are looking for HA in any CI/CD solution architecture we should then think about a **scalable, decoupled and flexible solution** that runs on a **resilient platform**. Today, these terms require us to think about containerized solutions orchestrated by [Kubernetes](https://kubernetes.io/).
+
+### A traditional Jenkins example
+
+[Jenkins](https://jenkins.io) is one of the most used solutions to orchestrate CI/CD pipelines. Its internal architecture is more than 10 years old, so we might think not to be the most advanced solution. But, the truth is that it has been evolving with these resilient best practices in mind (master-agent layers, pipeline decoupling and flexibility, containerized deployment, etc.), so if you deploy and design your platform and pipelines nicely you can get a true resilient CI/CD experience, meaning a CI/CD HA platform.
+
+But let's see the following picture:
+
+![Jenkins HA](/img/cicd-ha/CICD_HA_Archs.png)
+
+In a traditional deployment - like the diagram on the left - with Jenkins masters (orchestrators) and Jenkins agents (executors), we should at least work on an *Active-Passive* replication on masters to recover on time to provide a resilient platform with HA capabilities. But, if we also decouple the masters with smaller ones and deploy them natively in Kubernetes - diagram on the right -, the experience can be much more resilient without any replication method. A small master, that is running in a pod, can automatically be recovered by Kubernetes if it fails, restarting its service in a couple of minutes in the worse case. The pipeline could still be running by the way. Also, the rest of the masters are not impacted and agents are just ephemeral, only running and restarted when needed. So, service loss is not perceived.
```

Comment:
I would remove the commas in "A small master, that is running in a pod, can automatically"